File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static class Point {
        int x;
        int y;
        int depth;
        
        public Point(int x, int y, int depth) {
            this.x = x;
            this.y = y;
            this.depth = depth;
        }
    }
    
    static int dx[] = {0, 0, 1, -1};
    static int dy[] = {1, -1, 0, 0};
    public int solution(int[][] maps) {
        int answer = 0;
        int n = maps.length;
        int m = maps[0].length;
        
        int visited[][] = new int[n][m];
        
        for(int i = 0;i < n;i++) {
            Arrays.fill(visited[i], -1);
        }
        
        Queue<Point> q = new LinkedList<>();
        
        q.add(new Point(0, 0, 1));
        
        while(!q.isEmpty()) {
            Point temp = q.poll();
            
            for(int i = 0;i < 4;i++) {
                int nx = temp.x + dx[i];
                int ny = temp.y + dy[i];
                
                if(nx < 0 || ny < 0 || nx >= n || ny >= m || maps[nx][ny] == 0 || visited[nx][ny] != -1) continue;
                
                visited[nx][ny] = temp.depth + 1;
                
                q.add(new Point(nx, ny, visited[nx][ny]));
            }
        }
        
        return visited[n - 1][m - 1];
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var answer = 0;
    var n = maps.length, m = maps[0].length;
    var dx = [1, -1, 0, 0], dy = [0, 0, 1, -1];
    const visited = Array(n).fill(0).map(() => Array(m).fill(0));
    const q = [];
    q.push([0, 0]);
    visited[0][0] = 1;
    
    // 도착지의 상,좌가 막혀있으면 -1 리턴
    if(maps[n - 1][m - 2] === 0 && maps[n - 2][m - 1] === 0) return -1;
    
    while(q.length) {
        const [y, x] = q.shift();
        
        for(let i = 0;i < 4;i++) {
            let nx = x + dx[i];
            let ny = y + dy[i];
            
            if(ny < 0 || nx < 0 || ny >= n || nx >= m || maps[ny][nx] === 0 || visited[ny][nx]) continue;
            
            q.push([ny, nx]);
            visited[ny][nx] = visited[y][x] + 1;
        }
    }
    
    result = visited[n - 1][m - 1];
    
    if(!result) return -1;
    else return result;
    
    return answer;
}